flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -405,54 +404,30 @@ public void addMasterState(MasterState state) {
 	// ------------------------------------------------------------------------
 
 	/**
-	 * Aborts a checkpoint because it expired (took too long).
+	 * Aborts a checkpoint with reason and cause.
 	 */
-	public void abortExpired() {
+	public void abort(CheckpointFailureReason reason, Throwable cause) {
 		try {
-			Exception cause = new Exception("Checkpoint expired before completing");
-			onCompletionPromise.completeExceptionally(cause);
-			reportFailedCheckpoint(cause);
+			CheckpointException exception = new CheckpointException(reason, cause);
+			onCompletionPromise.completeExceptionally(exception);
+			reportFailedCheckpoint(exception);
+			assertAbortSubsumedForced(reason);
 		} finally {
 			dispose(true);
 		}
 	}
 
 	/**
-	 * Aborts the pending checkpoint because a newer completed checkpoint subsumed it.
+	 * Aborts a checkpoint with reason and cause.
 	 */
-	public void abortSubsumed() {
-		try {
-			Exception cause = new Exception("Checkpoints has been subsumed");
-			onCompletionPromise.completeExceptionally(cause);
-			reportFailedCheckpoint(cause);
-
-			if (props.forceCheckpoint()) {
-				throw new IllegalStateException("Bug: forced checkpoints must never be subsumed");
-			}
-		} finally {
-			dispose(true);
-		}
-	}
-
-
-	public void abortDeclined() {
-		abortWithCause(new Exception("Checkpoint was declined (tasks not ready)"));
-	}
-
-	/**
-	 * Aborts the pending checkpoint due to an error.
-	 * @param cause The error's exception.
-	 */
-	public void abortError(@Nonnull Throwable cause) {
-		abortWithCause(new Exception("Checkpoint failed: " + cause.getMessage(), cause));
+	public void abort(CheckpointFailureReason reason) {
+		abort(reason, null);
 	}
 
-	private void abortWithCause(@Nonnull Exception cause) {
-		try {
-			onCompletionPromise.completeExceptionally(cause);
-			reportFailedCheckpoint(cause);
-		} finally {
-			dispose(true);
+	private void assertAbortSubsumedForced(CheckpointFailureReason reason) {
+		if (props.forceCheckpoint()) {
```

Comment:
This looks incorrect, shouldn't it be `if (props.forceCheckpoint() && reason == CHECKPOINT_SUBSUMED)` ? I wonder why we also should not rather to this check in `CheckpointCoordinator#dropSubsumedCheckpoints`? It seems not needed to run this for every reason because it only applies to the subsumed case, which is already a distict path in the coordinator.

Reply:
Actually, I agree with the first part of this suggestion. For the second part, if we assume the behavior of the caller end, I think you are right. However, it seems a good experience is that we should not assume the caller's behavior, because it may be changed in the future, right? IMO, We can only assume someone calls the abort method, it could be `CheckpointCoordinator#dropSubsumedCheckpoints` or somewhere in a test case. So I think this method is valuable.